glambie/data/read_and_edit_bucket_csv_files.py

```diff
@@ -0,0 +1,343 @@
+    """
+    Function to generate a dataframe sumamrising which of the files need editing, and what edits need to be made for
+    these. Thought about doing the edits at the same time, but wanted to retain a record separate to the files
+    themselves of what changes have been made.
+
+    Parameters
+    ----------
+    downloaded_files : list[str]
+        List of files that have been downloaded to the local directory.
+    local_data_directory_path : str
+        Path where local copies of bucket are saved.
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe containing results of checks for inconsistencies within the csv files compared to the expected glambie
+        standard.
+    """
+
+    results_dict = {'local_filepath': downloaded_files, 'file_name': [os.path.basename(a) for a in downloaded_files]}
+    results_dataframe = pd.DataFrame.from_dict(results_dict)
+
+    results_dataframe['date_check_satisfied'] = np.nan
+    results_dataframe['nodata_check_satisfied'] = np.nan
+
+    for file in downloaded_files:
+        file_check_dataframe = check_glambie_submission_for_errors(file, results_dataframe)
+
+    file_check_dataframe.to_csv(os.path.join(local_path, 'record_of_edited_files.csv'))
+
+    return file_check_dataframe
+
+
+def check_glambie_submission_for_errors(csv_file_path: str, file_check_dataframe: pd.DataFrame) -> pd.DataFrame:
+    """
+    Function to perform consistency checks on a submitted glambie csv file. Currently performs 2 checks:
+
+    1. check for any non equal end_dates and subsequent start_dates: in some cases we see 1 day gaps between these
+    instead of overlapping dates as expected.
+    2. check for any instances of extreme values of change - we are aware of instances where people have used
+    e.g change = 9999. for rows where they don't have a measured change.
+
+    This function doesn't actually edit any files - instead it returns a record of which files should be edited and
+    why - which we want to save to retain full transparency for glambie participants.
+
+    Parameters
+    ----------
+    csv_file_path : str
+        Path to downloaded glambie csv file.
+    file_check_dataframe : pd.DataFrame
+        Dataframe containing results of checks for inconsistencies within the csv files compared to the expected glambie
+        standard.
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe containing input dataframe updated with the results of the checks for one file.
+    """
+    print(os.path.basename(csv_file_path))
+    submission_data_frame = pd.read_csv(csv_file_path)
+
+    # First, check for any non equal end_dates and subsequent start_dates
+    start_dates = submission_data_frame.start_date.values
+    end_dates = submission_data_frame.end_date.values
+    start_dates = np.append(start_dates, np.nan)  # Add an extra element to the end of this list for check below
+
+    date_check_bool = True
+    nodata_check_bool = True
```

Comment:
Likewise, I'm not 100% clear on this one. It looks like values outside of a range need to be set to no-data? Something like this?
```suggestion
    all_change_values_are_valid = True
```

Reply:
Here we're checking that submitters haven't decided to set a random no data value. Nans or no data rows are not allowed to be included in the submission, the row should just be dropped instead. So we're trying to remove extreme values that end up in the timeseries, which aren't actual measurements.